File: n09_admin/audit.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from hashlib import sha256
import json
from typing import Any, Mapping
from uuid import UUID, uuid4


FORBIDDEN_FIELD_PARTS = (
    "password",
    "secret",
    "token",
    "session_id",
    "authorization",
    "credential",
)


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


def _assert_safe(value: Any, path: str = "event") -> None:
    if isinstance(value, Mapping):
        for key, child in value.items():
            normalized = str(key).lower()
            if any(part in normalized for part in FORBIDDEN_FIELD_PARTS):
                raise ValueError(f"forbidden audit field: {path}.{key}")
            _assert_safe(child, f"{path}.{key}")
    elif isinstance(value, (list, tuple, set, frozenset)):
        for index, child in enumerate(value):
            _assert_safe(child, f"{path}[{index}]")


def _json_value(value: Any) -> Any:
    if isinstance(value, UUID):
        return str(value)
    if isinstance(value, datetime):
        return value.astimezone(timezone.utc).isoformat()
    if isinstance(value, (set, frozenset, tuple)):
        return sorted(_json_value(item) for item in value)
    if isinstance(value, Mapping):
        return {str(key): _json_value(child) for key, child in value.items()}
    return value


@dataclass(frozen=True, slots=True)
class AuditEvent:
    action: str
    result: str
    source: str
    correlation_id: UUID
    actor_id: UUID | None = None
    subject_id: UUID | None = None
    application_id: str | None = None
    cause: str | None = None
    role_id: str | None = None
    scope_type: str | None = None
    scope_id: str | None = None
    conditions: frozenset[str] = field(default_factory=frozenset)
    previous_value: Mapping[str, Any] | None = None
    new_value: Mapping[str, Any] | None = None
    justification: str = ""
    event_id: UUID = field(default_factory=uuid4)
    occurred_at: datetime = field(default_factory=_utc_now)

    def __post_init__(self) -> None:
        if self.occurred_at.tzinfo is None:
            raise ValueError("occurred_at must be timezone-aware")
        if not self.action.strip() or not self.result.strip() or not self.source.strip():
            raise ValueError("action, result and source are required")
        _assert_safe(self.previous_value)
        _assert_safe(self.new_value)

    def payload(self) -> dict[str, Any]:
        return _json_value(asdict(self))

    def canonical_json(self) -> str:
        return json.dumps(
            self.payload(), ensure_ascii=False, sort_keys=True, separators=(",", ":")
        )


def event_hash(event: AuditEvent, previous_hash: str) -> str:
    material = f"{previous_hash}\n{event.canonical_json()}".encode("utf-8")
    return sha256(material).hexdigest()

File: service-node/audit.mjs
import { createHash, randomUUID } from "node:crypto";

const FORBIDDEN_FIELD_PARTS = [
  "password",
  "secret",
  "token",
  "session_id",
  "authorization",
  "credential",
];

function assertSafe(value, path = "event") {
  if (Array.isArray(value)) {
    value.forEach((child, index) => assertSafe(child, `${path}[${index}]`));
    return;
  }
  if (!value || typeof value !== "object") return;
  for (const [key, child] of Object.entries(value)) {
    const normalized = key.toLowerCase();
    if (FORBIDDEN_FIELD_PARTS.some((part) => normalized.includes(part))) {
      throw new Error(`forbidden audit field: ${path}.${key}`);
    }
    assertSafe(child, `${path}.${key}`);
  }
}

function canonicalValue(value) {
  if (value instanceof Date) {
    return value.toISOString().replace(/\.(\d{3})Z$/, (_, milliseconds) =>
      milliseconds === "000" ? "+00:00" : `.${milliseconds}000+00:00`,
    );
  }
  if (Array.isArray(value)) return value.map(canonicalValue);
  if (value instanceof Set) return [...value].map(canonicalValue).sort();
  if (!value || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.keys(value).sort().map((key) => [key, canonicalValue(value[key])]),
  );
}

function deepFreeze(value) {
  if (!value || typeof value !== "object" || Object.isFrozen(value)) return value;
  Object.values(value).forEach(deepFreeze);
  return Object.freeze(value);
}

export function canonicalJson(value) {
  return JSON.stringify(canonicalValue(value));
}

export function createAuditEvent({
  action,
  result,
  source,
  correlationId,
  actorId = null,
  subjectId = null,
  applicationId = null,
  cause = null,
  roleId = null,
  scopeType = null,
  scopeId = null,
  conditions = [],
  previousValue = null,
  newValue = null,
  justification = "",
  eventId = randomUUID(),
  occurredAt = new Date(),
}) {
  if (![action, result, source, correlationId].every((item) => typeof item === "string" && item.trim())) {
    throw new Error("action, result, source and correlationId are required");
  }
  const event = {
    action: action.trim(),
    actor_id: actorId,
    application_id: applicationId,
    cause,
    conditions: [...conditions].map(canonicalValue).sort(),
    correlation_id: correlationId,
    event_id: eventId,
    justification: justification.trim(),
    new_value: canonicalValue(newValue),
    occurred_at: canonicalValue(occurredAt),
    previous_value: canonicalValue(previousValue),
    result: result.trim(),
    role_id: roleId,
    scope_id: scopeId,
    scope_type: scopeType,
    source: source.trim(),
    subject_id: subjectId,
  };
  assertSafe(event);
  return deepFreeze(event);
}

export function eventHash(event, previousHash = "") {
  return createHash("sha256")
    .update(`${previousHash}\n${canonicalJson(event)}`, "utf8")
    .digest("hex");
}

export function verifyAuditChain(entries) {
  let previousHash = "";
  for (const entry of entries) {
    if (entry.previousHash !== previousHash) return false;
    if (entry.eventHash !== eventHash(entry.event, previousHash)) return false;
    previousHash = entry.eventHash;
  }
  return true;
}
